sotodlib/preprocess/preprocess_util.py

```diff
@@ -0,0 +1,506 @@
+    """Storage policy assistance.  Helps to determine the HDF5
+    filename and dataset name for a result.
+
+    Make me better!
+
+    """
+    @staticmethod
+    def from_params(params):
+        if params['type'] == 'simple':
+            return ArchivePolicy(**params)
+        if params['type'] == 'directory':
+            return DirectoryArchivePolicy(**params)
+        raise ValueError('No handler for "type"="%s"' % params['type'])
+
+    def __init__(self, **kwargs):
+        self.filename = kwargs['filename']
+
+    def get_dest(self, product_id):
+        """Returns (hdf_filename, dataset_addr).
+
+        """
+        return self.filename, product_id
+
+
+class DirectoryArchivePolicy:
+    """Storage policy for stuff organized directly on the filesystem.
+
+    """
+    def __init__(self, **kwargs):
+        self.root_dir = kwargs['root_dir']
+        self.pattern = kwargs['pattern']
+
+    def get_dest(self, **kw):
+        """Returns full path to destination directory.
+
+        """
+        return os.path.join(self.root_dir, self.pattern.format(**kw))
+
+
+class _ReltimeFormatter(logging.Formatter):
+    def __init__(self, *args, t0=None, **kw):
+        super().__init__(*args, **kw)
+        if t0 is None:
+            t0 = time.time()
+        self.start_time = t0
+
+    def formatTime(self, record, datefmt=None):
+        if datefmt is None:
+            datefmt = '%8.3f'
+        return datefmt % (record.created - self.start_time)
+
+
+def init_logger(name, announce='', verbosity=2):
+    """Configure and return a logger for site_pipeline elements.  It is
+    disconnected from general sotodlib (propagate=False) and displays
+    relative instead of absolute timestamps.
+
+    """
+    logger = logging.getLogger(name)
+
+    if verbosity == 0:
+        level = logging.ERROR
+    elif verbosity == 1:
+        level = logging.WARNING
+    elif verbosity == 2:
+        level = logging.INFO
+    elif verbosity == 3:
+        level = logging.DEBUG
+
+    # add handler only if it doesn't exist
+    if len(logger.handlers) == 0:
+        ch = logging.StreamHandler(sys.stdout)
+        formatter = _ReltimeFormatter('%(asctime)s: %(message)s (%(levelname)s)')
+
+        ch.setLevel(level)
+        ch.setFormatter(formatter)
+        logger.addHandler(ch)
+        i, r = formatter.start_time // 1, formatter.start_time % 1
+        text = (time.strftime('%Y-%m-%d %H:%M:%S', time.gmtime(i))
+              + (',%03d' % (r*1000)))
+        logger.info(f'{announce}Log timestamps are relative to {text}')
+    else:
+        for handler in logger.handlers:
+            if isinstance(handler, logging.StreamHandler):
+                handler.setLevel(level)
+                break
+
+    logger.propagate = False
+    logger.setLevel(logging.DEBUG)
+
+    return logger
+
+
+def get_preprocess_context(configs, context=None):
+    """Load the provided config file and context file. To be used in
+    ``preprocess_*.py`` site pipeline scripts.
```

Comment:
I've had a number of people asking about using this function lately so I think its worth adding to the docstring that the key difference between using this function and just `cfg = yaml.safe_load(open(configs, 'r'))`, `context = core.Context(cfg['context_file'])` is that if the provided `context.yaml` does not have a metadata entry for preprocess then it will add one based on the definition in the `preprocess.yaml` file.